### ISSUE TITLE
revisit python discussion

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#1730]
I think it may be time that we discuss [python exclusion](https://github.com/coreos/fedora-coreos-tracker/blob/main/Design.md#approach-towards-shipping-python) again.

To be clear here:

1. I'm suggesting we no longer deny python inclusion
2. I'm not actually advocating for including python, but for being able to include tools that require python.
    - For example, if we somehow in the future removed all tools that require python, or existing tools get rewritten to not require python, then python could drop out at that point (since nothing we include would need it any longer)

Now, why am I suggesting this? I argue that excluding python has caused us more work/pain than if we had just included it.

- Package splitting.. We did manage to get some packages split out so that python deps were isolated to a sub package, but that was a lot of work and there are still some that aren't right:
    - The [nfs-utils](https://github.com/coreos/fedora-coreos-tracker/issues/121) saga is [still not completely sorted out](https://github.com/coreos/fedora-coreos-tracker/issues/572)
- We have had to keep out things that I think represent basic level functionality
    - like [firewalld](https://github.com/coreos/fedora-coreos-tracker/issues/26#issuecomment-415161012)
    - and [basic level SELinux tools](https://github.com/coreos/fedora-coreos-tracker/issues/126) (like `semanage`) that people expects to be there (https://github.com/coreos/fedora-coreos-tracker/issues/126)
- Also integrating with things like `ansible` hasn't really been able to work well (maybe some people have figured it out??). I know we have Ignition and things are supposed to be 'immutable', but practically people need to make small tweaks periodically.
- fwupd, which we [shipped not too long ago](https://github.com/coreos/fedora-coreos-tracker/issues/449) [appears broken](https://github.com/coreos/fedora-coreos-tracker/issues/1623#issuecomment-1851904401) because it dynamically depends on some python libraries to exist. I'm not saying we do want to include those libraries, just that we can't even consider it right now.

Now we could decide to implement a system python (where only files we ship in the OSTree could be executed with python), but we never made progress on investigating and implementing that IIUC. I think my suggestion is that we just stop resisting it and include it without new investigations to figure out if we can do something special here.

**Comment by LorbusChris** [comment]
Personally, I feel indifferent about the inclusion but I'd oppose the system python option, as it seems like the worst of both worlds to me (size increase && python still not universally usable).

**Comment by travier** [comment]
Related: https://github.com/coreos/fedora-coreos-tracker/issues/1122#issuecomment-2102779351

**Comment by cgwalters** [comment]
Also related is https://github.com/coreos/fedora-coreos-tracker/issues/1726 - when fcos derives from the base bootc image, then we'd pick up python by default.

> Also integrating with things like ansible hasn't really been able to work well 

Right, it is intentionally included there, see also https://docs.fedoraproject.org/en-US/bootc/dynamic-reconfiguration/

**Comment by travier** [comment]
See also: https://github.com/ostreedev/ostree/issues/1469

**Comment by dustymabe** [comment]
Another ancillary benefit if python were to get included is that we could take advantage of the new OSBuild buildroot support that was added in https://github.com/coreos/coreos-assembler/commit/0f8b35a3ba29b83337366c30000c423c6d3ace41 - of course this wouldn't be reason enough on it's own to include python, but would be a benefit if it were.

**Comment by c4rt0** [comment]
This was once more discussed today in the FCOS community meeting. 

We finally reached a consensus: 

> !agreed We will no longer exclude packages that require python from Fedora CoreOS. If we decide that we want the package in Fedora CoreOS we will include it, even if it's dependency chain includes python.

See the [full meeting log here](https://meetbot.fedoraproject.org/meeting-1_matrix_fedoraproject-org/2024-05-29/fedora-coreos-meeting.2024-05-29-16.32.log.html).

**Comment by dustymabe** [comment]
We discussed this in the community meeting today.

```
* INFO: `nfs-utils-coreos` dep chain includes `nfs-utils` again,
  which incudes `python`. We need to revisit the discussion in
  https://github.com/coreos/fedora-coreos-tracker/issues/1730 again.
  (@dustymabe:matrix.org, 16:55:00)
```

**Comment by travier** [comment]
We discussed this in today's community meeting, in the context of https://github.com/coreos/fedora-coreos-tracker/issues/1942:

```
AGREED: we'll wait a few weeks to see if a solution to the problem presents itself. At that point, if the issue hasn't been resolved, we'll drop the exclude for python for rawhide. If the problem still persists once we switch over to F43 for our stable streams then we'll keep python3 in the base in the future. Anytime before F43 it can be removed without issue if a solution to the problem exists. (@siosm:matrix.org, 17:06:48)
```

**Comment by dustymabe** [comment]
We are bumping up into the issue of not having python in our base again over in https://github.com/coreos/coreos-assembler/pull/4224

It's a bit of an implementation detail of OSBuild, but because OSBuild is python and runs in stages it requires python to be in the target "buildroot" to run.

For downstream we do have python in there so we can just use RHCOS/SCOS as the buildroot. For FCOS we don't have python in the payload so we use COSA as the build root. This leads to asymmetric upstream and downstream and also some hacks that we have to do.

Today this is the list of packages that get installed when layering in python3:

```
 Added:
  mpdecimal-4.0.1-2.fc43.x86_64
  python-pip-wheel-25.1.1-18.fc43.noarch
  python-unversioned-command-3.14.3-1.fc43.noarch
  python3-3.14.3-1.fc43.x86_64
  python3-libs-3.14.3-1.fc43.x86_64
```

**Comment by travier** [comment]
Another point here in favor of keeping Python out is that once we have it included, we will start missing packages adding a dependency on it, for good or (usually) bad reasons. So the image will grow without us noticing it.

**Comment by dustymabe** [comment]
> Another point here in favor of keeping Python out is that once we have it included, we will start missing packages adding a dependency on it, for good or (usually) bad reasons.

Agree that we would stop noticing when packages add a dep on python, but we'd also be able to stop chasing down and "fixing" things when that did happen. being the "python police" can be tiring. 

> So the image will grow without us noticing it.

Not sure I agree with this. Python would already be included so why would it grow? I guess if the package depended on not just python but also a python library we hadn't already included.

**Comment by dustymabe** [comment]
OK. One of the reasons I've been pushing for python is because of friction in our build system [1] where having python in the base OS would actually have made things easier. I have opened a PR against COSA https://github.com/coreos/coreos-assembler/pull/4534 to workaround this problem a different way. It's not as clean, but does mean we can move towards the benefits we desired without actually having python in the base OS.

[1] OSBuild is written in python and uses python for it's stages, which is perfectly normal, but coupled with our desire to not maintain a separate buildroot (we already have COSA and the OS itself to maintain) it means we'd have to manage a third thing.

**Comment by cgwalters** [comment]
In the general case we need to be able to generate minimal images without python, even if CoreOS adds it.

I think this is an osbuild design bug in general. 

> but coupled with our desire to not maintain a separate buildroot (we already have COSA and the OS itself to maintain)

But that said we already have a separate buildroot image that has python - the existing bootc base images. Why not just tell osbuild to use those, which should be in sync anyways?

I think centos automotive did something similar in this space btw.

**Comment by dustymabe** [comment]
> In the general case we need to be able to generate minimal images without python, even if CoreOS adds it.
> 
> I think this is an osbuild design bug in general.

yeah. It's certainly caused some friction on our end.

> 
> > but coupled with our desire to not maintain a separate buildroot (we already have COSA and the OS itself to maintain)
> 
> But that said we already have a separate buildroot image that has python - the existing bootc base images. Why not just tell osbuild to use those, which should be in sync anyways?

I hadn't thought about that. We are using the bootc base image already for composing (i.e. container build) our FCOS container image, but not disk images.

I still think I like https://github.com/coreos/coreos-assembler/pull/4534 better for now, until we find problems with it. Since it's basically a 1:1 match of the CoreOS payload + python. i.e. fasttracking `xfsprogs` in CoreOS will fix a problem that happened when building disk images in the buildroot. etc..

Will see what others think in the code review of that PR.

**Comment by cgwalters** [comment]
>  i.e. fasttracking xfsprogs in CoreOS 

Yes but we could fast track that in the bootc base image or in Fedora in general too right?

**Comment by dustymabe** [comment]
This is getting a bit off topic here, but..

> Yes but we could fast track that in the bootc base image or in Fedora in general too right?

I'm not aware of any mechanism to fast track things in Fedora outside of bodhi. Do we have the ability to do that in the bootc base images?

**Comment by cgwalters** [comment]
> I still think I like https://github.com/coreos/coreos-assembler/pull/4534 better for now,

What's happening in cosa there was quite inscrutable before and this is piling on to that though...

To me a big goal of the containerization initiative was trying to simplify things so that this could all be more approachable and understandable as a container build, and looking at stuff like this that has not been a success. Now I know there's not one reason or person at fault or anything, it's hard to change the engine as it's running of course.
